core/src/main/scala/datas.scala

```diff
@@ -203,43 +231,78 @@ object datas {
     def compileReference: Reference ~> TypedFragment
   }
 
-  final case class TypedFragment[Type](frag: Fragment, read: Read[Type]) {
-    def map[B](f: Type => B): TypedFragment[B] = copy(read = read.map(f))
+  sealed trait TypedFragment[Type] extends Product with Serializable {
+    def optional: TypedFragment[Option[Type]] = TypedFragment.Optional(this)
+    def map[B](f: Type => B): TypedFragment[B] = TypedFragment.Map(this, f)
+    def product[B](another: TypedFragment[B]): TypedFragment[(Type, B)] = TypedFragment.Product(this, another)
 
-    def product[B](another: TypedFragment[B]): TypedFragment[(Type, B)] = {
-      implicit val rl = read
-      implicit val rr = another.read
-      val _ = (rl, rr)
-      TypedFragment(frag ++ fr"," ++ another.frag, Read[(Type, B)])
+    def compile: (Fragment, Either[Get[Type], Read[Type]]) = TypedFragment.toFragAndGet(this)
+  }
+
+  object TypedFragment {
+
+    def toFragAndGet[Type]: TypedFragment[Type] => (Fragment, Either[Get[Type], Read[Type]]) = {
+      case Single(frag, get) => (frag, get.asLeft)
+      case Optional(underlying) =>
+        toFragAndGet(underlying).map {
+          case Left(get)   => Read.fromGetOption(get).asRight
+          case Right(read) => read.map(_.some).asRight
+        }
+      case p: Product[a, b] =>
+        def toRead[A]: Either[Get[A], Read[A]] => Read[A] = _.fold(Read.fromGet(_), identity)
+
+        val (leftFrag, leftRead) = toFragAndGet(p.l).map(toRead)
+        val (rightFrag, rightRead) = toFragAndGet(p.r).map(toRead)
+
+        implicit val lR = leftRead
+        implicit val rR = rightRead
+
+        (leftFrag ++ fr", " ++ rightFrag, Read[(a, b)].asRight)
+      case Map(underlying, f) => toFragAndGet(underlying).map(_.bimap(_.map(f), _.map(f)))
     }
+    final case class Single[Type](frag: Fragment, get: Get[Type]) extends TypedFragment[Type]
+    final case class Optional[Type](underlying: TypedFragment[Type]) extends TypedFragment[Option[Type]]
```

Comment:
Possibly at this point we could ditch this and do the optioning on the level above.